src/Raven.Server/Documents/HugeDocuments.cs

```diff
@@ -1,35 +1,95 @@
+        private readonly NotificationCenter.NotificationCenter _notificationCenter;
+        private readonly NotificationsStorage _notificationsStorage;
+        private readonly string _database;
 
-        public HugeDocuments(int maxCollectionSize, long maxWarnSize)
+        public HugeDocuments(NotificationCenter.NotificationCenter notificationCenter, NotificationsStorage notificationsStorage, string database, int maxCollectionSize, long maxWarnSize)
         {
+            _notificationCenter = notificationCenter;
+            _notificationsStorage = notificationsStorage;
+            _database = database;
             _maxWarnSize = maxWarnSize;
             _hugeDocs = new SizeLimitedConcurrentDictionary<Tuple<string, DateTime>, long>(maxCollectionSize);
         }
 
         public void AddIfDocIsHuge(Document doc)
         {
-            if(doc.Data.Size > _maxWarnSize)
+            if (doc.Data.Size > _maxWarnSize)
+            {
                 _hugeDocs.Set(new Tuple<string, DateTime>(doc.Id, DateTime.UtcNow), doc.Data.Size);
-
+                AddHint(doc.Id, doc.Data.Size);
+            }
         }
 
         public void AddIfDocIsHuge(string id, int size)
         {
             if (size > _maxWarnSize)
+            {
                 _hugeDocs.Set(new Tuple<string, DateTime>(id, DateTime.UtcNow), size);
+                AddHint(id, size);
+            }
         }
 
+        private void AddHint(string id, int size)
+        {
+            lock (_addHintSyncObj)
+            {
+                var performanceHint = GetOrCreatePerformanceHint(out var details);
+                details.Update(id, size);
+                _notificationCenter.Add(performanceHint);
+            }
+        }
+        
         public SizeLimitedConcurrentDictionary<Tuple<string, DateTime>, long> GetHugeDocuments()
         {
             return _hugeDocs;
         } 
+        
+        private PerformanceHint GetOrCreatePerformanceHint(out HugeDocumentsDetails details)
+        {
+            //Read() is transactional, so this is thread-safe
+            using (_notificationsStorage.Read(HugeDocumentsId, out var ntv))
+            {
+                if (ntv == null || ntv.Json.TryGet(nameof(PerformanceHint.Details), out BlittableJsonReaderObject detailsJson) == false || detailsJson == null)
+                {
+                    details = new HugeDocumentsDetails();
+                }
+                else
+                {
+                    details = (HugeDocumentsDetails)EntityToBlittable.ConvertToEntity(
+                        typeof(HugeDocumentsDetails),
+                        HugeDocumentsId,
+                        detailsJson,
+                        DocumentConventions.Default);
+                }
+
+                return PerformanceHint.Create(
+                    _database,
+                    "Huge documents",
+                    "We have detected that some documents has surpassed the configured size threshold",
```

Comment:
We need a better message here:
- explain that huge docs might impact performance
- add the current threshold
- provide configuration key name

Reply:
updated.